"No value available for annotation member" for MQTT and Micronaut 2.4.0-SNAPSHOT

The MQTT GraalVM app has detected an error when starting the MQTT app with Micronaut 2.4.0-SNAPSHOT. It happens on both JVM and GraalVM.

The error seems to be in `io.micronaut.core.annotation.AnnotationValue.lambda$getRequiredValue$1(AnnotationValue.java:774)` that's why I'm opening the issue here instead of in micronaut-mqtt.

### Steps to Reproduce

- `git clone https://github.com/micronaut-graal-tests/micronaut-mqtt-graal`
- `cd micronaut-mqtt-graal`
- `git checkout 2.4.x_v3`
- Start MQTT docker container as explained in the README: `docker run -it --rm -p 1883:1883 eclipse-mosquitto:1.6.12`
- `./gradlew run`

### Expected Behaviour

The application should start

### Actual Behaviour

The application fails with error:

```
 __  __ _                                  _   
|  \/  (_) ___ _ __ ___  _ __   __ _ _   _| |_ 
| |\/| | |/ __| '__/ _ \| '_ \ / _` | | | | __|
| |  | | | (__| | | (_) | | | | (_| | |_| | |_ 
|_|  |_|_|\___|_|  \___/|_| |_|\__,_|\__,_|\__|
  Micronaut (v2.4.0-SNAPSHOT)

14:27:49.811 [main] ERROR i.m.m.e.DefaultMqttSubscriberExceptionHandler - Subscriber [void receiveMessage(MessageRequest message)] encountered an error
io.micronaut.mqtt.exception.MqttSubscriberException: Subscriber [void receiveMessage(MessageRequest message)] encountered an error
        at io.micronaut.mqtt.intercept.AbstractMqttSubscriberAdvice.handleException(AbstractMqttSubscriberAdvice.java:128)
        at io.micronaut.mqtt.intercept.AbstractMqttSubscriberAdvice.process(AbstractMqttSubscriberAdvice.java:112)
        at io.micronaut.mqtt.intercept.AbstractMqttSubscriberAdvice.process(AbstractMqttSubscriberAdvice.java:54)
        at io.micronaut.context.AnnotationProcessorListener.onCreated(AnnotationProcessorListener.java:95)
        at io.micronaut.context.AnnotationProcessorListener.onCreated(AnnotationProcessorListener.java:44)
        at io.micronaut.context.DefaultBeanContext.doCreateBean(DefaultBeanContext.java:2000)
        at io.micronaut.context.DefaultBeanContext.addCandidateToList(DefaultBeanContext.java:3091)
        at io.micronaut.context.DefaultBeanContext.getBeanRegistrations(DefaultBeanContext.java:2968)
        at io.micronaut.context.DefaultBeanContext.streamOfType(DefaultBeanContext.java:783)
        at io.micronaut.context.DefaultBeanContext.streamOfType(DefaultBeanContext.java:770)
        at io.micronaut.context.DefaultBeanContext.lambda$initializeContext$32(DefaultBeanContext.java:1612)
        at java.util.HashMap.forEach(HashMap.java:1289)
        at io.micronaut.context.DefaultBeanContext.initializeContext(DefaultBeanContext.java:1611)
        at io.micronaut.context.DefaultApplicationContext.initializeContext(DefaultApplicationContext.java:219)
        at io.micronaut.context.DefaultBeanContext.readAllBeanDefinitionClasses(DefaultBeanContext.java:2857)
        at io.micronaut.context.DefaultBeanContext.start(DefaultBeanContext.java:231)
        at io.micronaut.context.DefaultApplicationContext.start(DefaultApplicationContext.java:165)
        at io.micronaut.runtime.Micronaut.start(Micronaut.java:71)
        at io.micronaut.runtime.Micronaut.run(Micronaut.java:311)
        at io.micronaut.runtime.Micronaut.run(Micronaut.java:297)
        at example.micronaut.Application.main(Application.java:8)
Caused by: java.lang.IllegalStateException: No value available for annotation member @io.micronaut.mqtt.annotation.Topic[qos] of type: int
        at io.micronaut.core.annotation.AnnotationValue.lambda$getRequiredValue$1(AnnotationValue.java:774)
        at java.util.Optional.orElseThrow(Optional.java:290)
        at io.micronaut.core.annotation.AnnotationValue.getRequiredValue(AnnotationValue.java:774)
        at io.micronaut.mqtt.intercept.AbstractMqttSubscriberAdvice.process(AbstractMqttSubscriberAdvice.java:91)
        ... 19 common frames omitted
```

### Environment Information

- **Operating System**: Linux Mint 20.1
- **Micronaut Version:** 2.4.0-SNAPSHOT
- **JDK Version:** 8

--- a/inject-java/src/test/groovy/io/micronaut/inject/annotation/AnnotationDefaultValuesSpec.groovy
+++ b/inject-java/src/test/groovy/io/micronaut/inject/annotation/AnnotationDefaultValuesSpec.groovy
@@ -1,10 +1,32 @@
 package io.micronaut.inject.annotation
 
 import io.micronaut.context.annotation.Parameter
+import io.micronaut.core.annotation.AnnotationValue
 import io.micronaut.inject.AbstractTypeElementSpec
 import io.micronaut.inject.BeanDefinition
+import io.micronaut.inject.annotation.repeatable.Topic
 
 class AnnotationDefaultValuesSpec extends AbstractTypeElementSpec {
+
+    void 'test getRequiredValue from AnnotationValue'() {
+        given:
+        BeanDefinition definition = buildBeanDefinition('issue5048.Test','''\
+package issue5048;
+
+import io.micronaut.inject.annotation.repeatable.Topic;
+
+@javax.inject.Singleton
+@Topic("test")
+class Test {
+    
+}
+''')
+        when:
+        def annotationValue = definition.getAnnotationValuesByType(Topic)[0]
+        then:
+        annotationValue.getRequiredValue("qos", Integer.class) == 1
+
+    }
 
     void "test write annotation default values for constructor arguments"() {
         given:

--- a/inject-java/src/test/groovy/io/micronaut/inject/annotation/repeatable/Topic.java
+++ b/inject-java/src/test/groovy/io/micronaut/inject/annotation/repeatable/Topic.java
@@ -1,0 +1,13 @@
+package io.micronaut.inject.annotation.repeatable;
+
+import java.lang.annotation.Repeatable;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+@Retention(RetentionPolicy.RUNTIME)
+@Repeatable(Topics.class)
+public @interface Topic {
+    String value();
+
+    int qos() default 1;
+}

--- a/inject-java/src/test/groovy/io/micronaut/inject/annotation/repeatable/Topics.java
+++ b/inject-java/src/test/groovy/io/micronaut/inject/annotation/repeatable/Topics.java
@@ -1,0 +1,9 @@
+package io.micronaut.inject.annotation.repeatable;
+
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+
+@Retention(RetentionPolicy.RUNTIME)
+public @interface Topics {
+    Topic[] value();
+}

--- a/inject/src/main/java/io/micronaut/inject/annotation/AnnotationMetadataSupport.java
+++ b/inject/src/main/java/io/micronaut/inject/annotation/AnnotationMetadataSupport.java
@@ -44,7 +44,7 @@ import java.util.function.Function;
  * @since 1.0
  */
 @Internal
-class AnnotationMetadataSupport {
+public final class AnnotationMetadataSupport {
 
     private static final Map<String, Map<String, Object>> ANNOTATION_DEFAULTS = new ConcurrentHashMap<>(20);
 
@@ -92,7 +92,8 @@ class AnnotationMetadataSupport {
      * @param annotation The annotation
      * @return The default values for the annotation
      */
-    static Map<String, Object> getDefaultValues(String annotation) {
+    @UsedByGeneratedCode
+    public static Map<String, Object> getDefaultValues(String annotation) {
         return ANNOTATION_DEFAULTS.computeIfAbsent(annotation, s -> Collections.emptyMap());
     }
 

--- a/inject/src/main/java/io/micronaut/inject/annotation/AnnotationMetadataWriter.java
+++ b/inject/src/main/java/io/micronaut/inject/annotation/AnnotationMetadataWriter.java
@@ -77,6 +77,14 @@ public class AnnotationMetadataWriter extends AbstractClassFileWriter {
             )
     );
 
+    private static final org.objectweb.asm.commons.Method METHOD_GET_DEFAULT_VALUES = org.objectweb.asm.commons.Method.getMethod(
+            ReflectionUtils.getRequiredInternalMethod(
+                    AnnotationMetadataSupport.class,
+                    "getDefaultValues",
+                    String.class
+            )
+    );
+
     private static final org.objectweb.asm.commons.Method CONSTRUCTOR_ANNOTATION_METADATA = org.objectweb.asm.commons.Method.getMethod(
             ReflectionUtils.getRequiredInternalConstructor(
                     DefaultAnnotationMetadata.class,
@@ -96,17 +104,11 @@ public class AnnotationMetadataWriter extends AbstractClassFileWriter {
             )
     );
 
-    private static final org.objectweb.asm.commons.Method CONSTRUCTOR_ANNOTATION_VALUE = org.objectweb.asm.commons.Method.getMethod(
-            ReflectionUtils.getRequiredInternalConstructor(
-                    io.micronaut.core.annotation.AnnotationValue.class,
-                    String.class
-            )
-    );
-
     private static final org.objectweb.asm.commons.Method CONSTRUCTOR_ANNOTATION_VALUE_AND_MAP = org.objectweb.asm.commons.Method.getMethod(
             ReflectionUtils.getRequiredInternalConstructor(
                     io.micronaut.core.annotation.AnnotationValue.class,
                     String.class,
+                    Map.class,
                     Map.class
             )
     );
@@ -613,10 +615,12 @@ public class AnnotationMetadataWriter extends AbstractClassFileWriter {
 
             if (CollectionUtils.isNotEmpty(values)) {
                 pushAnnotationAttributes(declaringType, declaringClassWriter, methodVisitor, values, loadTypeMethods);
-                methodVisitor.invokeConstructor(annotationValueType, CONSTRUCTOR_ANNOTATION_VALUE_AND_MAP);
             } else {
-                methodVisitor.invokeConstructor(annotationValueType, CONSTRUCTOR_ANNOTATION_VALUE);
+                methodVisitor.getStatic(Type.getType(Collections.class), "EMPTY_MAP", Type.getType(Map.class));
             }
+            methodVisitor.push(annotationName);
+            methodVisitor.invokeStatic(Type.getType(AnnotationMetadataSupport.class), METHOD_GET_DEFAULT_VALUES);
+            methodVisitor.invokeConstructor(annotationValueType, CONSTRUCTOR_ANNOTATION_VALUE_AND_MAP);
         } else {
             methodVisitor.visitInsn(ACONST_NULL);
         }

